### PR TITLE
[WIP] merge sds change on cluster to master

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/6411

cherrypick PRs from https://github.com/istio/istio/issues/6411 about sds change on cluster

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -540,7 +574,7 @@ func buildDefaultCluster(env *model.Environment, name string, discoveryType v2.C
 	}
 
 	defaultTrafficPolicy := buildDefaultTrafficPolicy(env, discoveryType)
-	applyTrafficPolicy(cluster, defaultTrafficPolicy, nil)
+	applyTrafficPolicy(cluster, defaultTrafficPolicy, nil, env.Mesh.SdsRefreshDelay, env.Mesh.SdsUdsPath)
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Isn't SDS push based ? Why does it have a refresh delay ?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -467,27 +480,48 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
...
-			},
-			Sni: tls.Sni,
+			}
+		} else {
+			saDomainCanonicalPrefix := fmt.Sprintf("%s://%s", spiffe, canonicalSAPrefix)
+			cluster.TlsContext.CommonTlsContext.ValidationContextType = model.ConstructValidationContext(model.CARootCertPath, tls.SubjectAltNames)
+			cluster.TlsContext.CommonTlsContext.TlsCertificateSdsSecretConfigs = []*auth.SdsSecretConfig{}
+			refreshDuration, _ := types.DurationFromProto(refreshDuration)
+
+			for _, sa := range tls.SubjectAltNames {
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Confused - how will this work for normal secrets or those without the 'canonical-service-accounts' annotation ? Why is SDS restricted to this ?

**Comment by costinm** [comment]
Agreed.

Env variables for temporary tweaks - anything long term or that user will
see in normal usage must be part of configuration.

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 1 above)
Comment (quanjielin) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: removed ```sdsrefreshdelay``` in https://github.com/istio/istio/pull/8843 and https://github.com/istio/api/pull/637

**Comment by quanjielin** [comment]
@costinm @rshriram https://github.com/istio/istio/pull/8879 to address the env variable comment

**Comment by quanjielin** [comment]
send out https://github.com/istio/istio/pull/9128 to resolve rest comments left in this PR, so close this one.

**Pull request closed without merging**